30-indoor-air-quality-and-gas-leak/firmware/sparrow-application/aqi/aqi.cpp

```diff
@@ -0,0 +1,496 @@
+    {
+        MX_I2C2_DeInit();
+    }
+
+    bool ping(uint8_t i2c_address)
+    {
+        return MY_I2C2_Ping(i2c_address, timeout_ms, retries);
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////////////////////
+    // writeRegisterByte()
+    //
+    // Write a byte to a register
+
+    bool writeRegisterByte(uint8_t i2c_address, uint8_t offset, uint8_t dataToWrite)
+    {
+        return MY_I2C2_WriteRegister(i2c_address, offset, &dataToWrite, 1, timeout_ms);
+    }
+
+    int writeRegisterRegion(uint8_t i2c_address, uint8_t offset, const uint8_t *data, uint16_t length)
+    {
+        return MY_I2C2_WriteRegister(i2c_address, offset, (void*)data, length, timeout_ms) ? 0 : -1;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // readRegisterRegion()
+    //
+    // Reads a block of data from an i2c register on the devices.
+    //
+    // For large buffers, the data is chuncked over KMaxI2CBufferLength at a time
+    //
+    //
+    int readRegisterRegion(uint8_t addr, uint8_t reg, uint8_t *data, uint16_t numBytes)
+    {
+        return MY_I2C2_ReadRegister(addr, reg, data, numBytes, timeout_ms) ? 0 : -1;
+    }
+};
+
+class AQISensor :  public QwDevENS160 {
+
+    STM32HALI2CBus bus;
+    
+public:
+    AQISensor():
+        bus(AQI_I2C_TIMEOUT_MS, AQI_I2C_RETRY_COUNT)
+    {
+        setCommunicationBus(bus, ENS160_ADDRESS_HIGH);
+    }
+
+    bool read(AQISensorReading& reading) {
+        bool success = false;
+        if (bus.init()) {
+            if (!this->setOperatingMode(SFE_ENS160_STANDARD)) {
+                APP_PRINTF("aqi: ERROR setting operating mode to standard\r\n");
+            }
+            else if (!this->checkDataStatus()) {
+                APP_PRINTF("aqi: ERROR checkDataStatus\r\n");
+            }
+            else {
+                reading.aqi = this->getAQI();
+                reading.eco2 = this->getECO2();
+                reading.tvoc = this->getTVOC();
+                reading.etoh = this->getETOH();
+                if (reading.aqi && reading.eco2) {
+                    success = true;
+                }
+                else {
+                    APP_PRINTF("aqi: ERROR reading is 0.\r\n");
+                }
+            }
+            //this->setOperatingMode(SFE_ENS160_IDLE);
+            bus.deinit();
+        }
+        else {
+
+        }
+        return success;
+    }
+};
+
+AQISensor aqiSensor;
+
+
+// Addresses of the BME sensor used to identify
+// the Sparrow Reference Sensor Board
+#define BME280_I2C_ADDR_PRIM        UINT8_C(0x76)
+#define BME280_I2C_ADDR_SEC         UINT8_C(0x77)
+#define BME280_I2C_RETRY_COUNT      3
+#define BME280_I2C_TIMEOUT_MS       100
+
+
+
+// We have no viable way of detecting whether or not the PIR sensor
+// hardware is present, so we use the presence of the BME280 as a proxy.
+bool isSparrowReferenceSensorBoard (void) {
```

Comment:
It seems like this isn't used anywhere.

Reply:
It isn't.  I tried using it but it failed.  I'll remove it for now. I think our detection of runtime mode not the best and you, Zak and I should chat about it next week. The app framework could  indicate to each app (ideally in appSchedInit() or at some later point) whether the device is running as a gateway or node.